src/Microsoft.DotNet.Interactive.Tests/LanguageServices/CompletionTests.cs

```diff
@@ -190,5 +192,66 @@ public async Task Completions_are_available_for_symbols_declared_in_a_submission
                              because: $"position {position} should provide completions");
             }
         }
+
+        [Theory]
+        [InlineData("[|#!t|]", "#!two,#!twilight", null, Skip = "Composite kernel magic command completions not working yet")]
+        [InlineData("[|#!tw|]\n", "#!two,#!twilight", null, Skip = "Composite kernel magic command completions not working yet")]
+        [InlineData("[|#!t|]", "#!two,#!twilight", Language.CSharp)]
+        [InlineData("[|#!tw|]\n", "#!two,#!twilight", Language.CSharp)]
+        [InlineData("[|#!t|]", "#!two,#!twilight", Language.FSharp)]
+        [InlineData("[|#!tw|]\n", "#!two,#!twilight", Language.FSharp)]
+        [InlineData("[|#!t|]", "#!two,#!twilight", Language.PowerShell)]
+        [InlineData("[|#!tw|]\n", "#!two,#!twilight", Language.PowerShell)]
+        public async Task Completions_are_available_for_magic_commands_added_at_runtime(
+            string markupCode,
+            string expected, 
+            Language? language)
+        {
+            MarkupTestFile.GetSpan(markupCode, out var code, out var span);
+
+            var sourceText = SourceText.From(code);
+
+            var kernel = CreateKernel();
+
+            using var _ = new AssertionScope();
+
+            _output.WriteLine($"Checking positions from {span.Start} to {span.End}");
+
+            KernelBase kernelToExtend = kernel;
+            if (language != null)
+            {
+                kernelToExtend = kernel.FindKernel(language.Value) as KernelBase;
+            }
+
+            kernelToExtend.AddDirective(new Command("#!two")
+            {
+                Handler = CommandHandler.Create(() => { })
+            });
+
+                kernelToExtend.AddDirective(new Command("#!twilight")
+            {
+                Handler = CommandHandler.Create(() => { })
+            });
+
+
+            foreach (var position in Enumerable.Range(span.Start, span.Length + 1))
```

Comment:
Rather than duplicate this, I wonder if we shouldn't extract a custom assertion. There are more completion tests to be written still, and we'll also want to add verifications for things like documentation and item kind.